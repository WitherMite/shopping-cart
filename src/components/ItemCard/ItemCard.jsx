import PropTypes from "prop-types";

const USDFormatter = Intl.NumberFormat("en-US", {
  style: "currency",
  currency: "USD",
});

ItemCard.propTypes = {
  handleClick: PropTypes.func,
  image: PropTypes.string,
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  price: PropTypes.string.isRequired,
};

export default function ItemCard({
  image,
  title,
  description,
  price,
  handleClick,
}) {
  return (
    <button onClick={() => handleClick(1)}>
      <img src={image} alt="" />
      <h2>{title}</h2>
      <p>{description}</p>
      <p>{USDFormatter.format(price)}</p>
    </button>
  );
}
