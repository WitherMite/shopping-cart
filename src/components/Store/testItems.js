import logo from "../../assets/logo-ipsum-small.svg";

const testItems = [
  {
    id: 1,
    title: "Item 1",
    price: "40",
    category: "test",
    description:
      "This is item number 1, there are many like it, but this is the first",
    image: logo,
  },
  {
    id: 2,
    title: "Item 2",
    price: "12",
    category: "test",
    description: "Lorem ipsum sit amet dolor...",
    image: logo,
  },
  {
    id: 3,
    title: "Item 3",
    price: "20.45",
    category: "test",
    description: "Three.",
    image: logo,
  },
];

export default testItems;
