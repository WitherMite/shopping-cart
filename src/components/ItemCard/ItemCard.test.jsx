import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ItemCard from "../ItemCard/ItemCard.jsx";
import testItems from "../Store/testItems.js";

// have description be hidden until element is clicked

describe("ItemCard component", () => {
  it("Renders item information", () => {
    const { container } = render(
      <ItemCard {...testItems[0]} handleClick={() => {}} />
    );
    expect(container).toMatchSnapshot();
  });

  it("Renders controlled input field for count", async () => {
    const user = userEvent.setup();
    render(<ItemCard {...testItems[1]} handleClick={() => {}} />);

    const input = screen.getByRole("spinbutton", { name: "Count:" });
    const initialVal = input.value;
    await user.click(input);
    await user.keyboard("{backspace}{4}");
    const secondVal = input.value;
    await user.click(input);
    await user.keyboard("{backspace}{1}{0}{0}");
    const thirdVal = input.value;

    expect(initialVal).toBe("1");
    expect(secondVal).toBe("4");
    expect(thirdVal).toBe("100");
  });

  it("Clicking button calls handleClick with input's number value", async () => {
    const user = userEvent.setup();
    const fn = vi.fn();
    render(<ItemCard {...testItems[2]} handleClick={fn} />);

    const button = screen.getByRole("button", { name: "Add to Cart" });
    await user.click(button);
    const input = screen.getByRole("spinbutton", { name: "Count:" });
    await user.click(input);
    await user.keyboard("{backspace}{4}");
    await user.click(button);

    expect(fn).toHaveBeenNthCalledWith(1, 1);
    expect(fn).toHaveBeenNthCalledWith(2, 4);
  });

  it("Clicking button resets count input's value", async () => {
    const user = userEvent.setup();
    render(<ItemCard {...testItems[1]} handleClick={() => {}} />);

    const input = screen.getByRole("spinbutton", { name: "Count:" });
    const initialVal = input.value;
    await user.click(input);
    await user.keyboard("{backspace}{4}");
    const secondVal = input.value;
    const button = screen.getByRole("button", { name: "Add to Cart" });
    await user.click(button);
    const thirdVal = input.value;

    expect(thirdVal).toBe(initialVal);
    expect(thirdVal).not.toBe(secondVal);
  });
});
