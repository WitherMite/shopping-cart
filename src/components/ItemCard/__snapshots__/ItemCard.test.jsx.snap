// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ItemCard component > Renders item information 1`] = `
<div>
  <section
    class="store-item-card"
  >
    <img
      alt=""
      class="item-img"
      src="/src/assets/logo-ipsum-small.svg"
    />
    <h2
      class="item-title"
    >
      Item 1
    </h2>
    <p
      class="item-description"
    >
      This is item number 1, there are many like it, but this is the first
    </p>
    <p
      class="item-price"
    >
      $40.00
    </p>
    <div
      class="item-count-field"
    >
      <label
        for="count-1"
      >
        Count:
      </label>
      <input
        id="count-1"
        min="1"
        step="1"
        type="number"
        value="1"
      />
    </div>
    <button
      class="add-item-btn"
    >
      Add to Cart
    </button>
  </section>
</div>
`;
